icu4c/source/i18n/unitconverter.cpp

```diff
@@ -5,14 +5,498 @@
 
 #if !UCONFIG_NO_FORMATTING
 
+#include <cmath>
+
+#include "charstr.h"
+#include "double-conversion.h"
+#include "measunit_impl.h"
+#include "resource.h"
+#include "uassert.h"
+#include "unicode/stringpiece.h"
+#include "unicode/unistr.h"
+#include "unicode/utypes.h"
+#include "unitconverter.h"
+#include "uresimp.h"
+
 U_NAMESPACE_BEGIN
 
+namespace {
+
+using icu::double_conversion::StringToDoubleConverter;
+
+/* Internal Structure */
+
+// Represents a raw convert unit.
+struct ConvertUnit {
+    StringPiece source;
+    StringPiece target;
+
+    // NOTE
+    //  If the source and target are equal, the factor will be "1"
+    StringPiece factor = "1";
+
+    // NOTE
+    //  If the offset is not exist, its value will be "0"
+    StringPiece offset = "0";
+
+    // NOTE
+    //  If reciprocal is not exits, its value will be `false`
+    bool reciprocal = false;
+};
+
+typedef enum Signal {
+    NEGATIVE = -1,
+    POSITIVE = 1,
+} Signal;
+
+enum UnitsCase {
+    RECIPROCAL,
+    CONVERTIBLE,
+    UNCONVERTIBLE,
+};
+
+/* Helpers */
+
+// Returns `double` from a scientific number(i.e. "1", "2.01" or "3.09E+4")
+double strToDouble(StringPiece strNum) {
+    std::string charNum;
```

Comment:
You can't use std::string in library code.  I think you can just delete charNum.

Reply:
Done !